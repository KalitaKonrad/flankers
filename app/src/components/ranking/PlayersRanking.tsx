import React, { useState } from 'react';
import { FlatList, ListRenderItemInfo, StyleSheet, View } from 'react-native';
import { List, Text, useTheme } from 'react-native-paper';

import { UserProfilePayload } from '../../types/userProfilePayload';
import { Avatar } from '../shared/Avatar';

interface PlayersRankingProps {
  players: UserProfilePayload[];
  pageNumber: number;
  userId: number;
  onEndReached(): void;
}

export const PlayersRanking: React.FC<PlayersRankingProps> = ({
  players,
  pageNumber,
  userId,
  onEndReached,
}) => {
  const theme = useTheme();

  const [onEndReachedState, setOnEndReachedState] = useState(false);

  const renderItem = ({
    item,
    index,
  }: ListRenderItemInfo<UserProfilePayload>) => (
    <List.Item
      title={item.name}
      titleStyle={styles.playerName}
      description={`${item.elo} punktÃ³w rankingowych`}
      left={() => (
        <View style={styles.playerAvatarContainer}>
          <Avatar size={40} borderRadius={8} src={{ uri: item.avatar }} />
        </View>
      )}
      right={() => (
        <Text style={styles.rightIndexStyle}>
          #{index + (pageNumber - 1) * 10 + 1}
        </Text>
      )}
      style={
        item.id === userId
          ? {
              backgroundColor: theme.colors.primary,
              borderRadius: 8,
              opacity: 0.5,
            }
          : { backgroundColor: theme.colors.white }
      }
    />
  );

  return (
    <FlatList
      data={players}
      renderItem={renderItem}
      contentContainerStyle={styles.container}
      keyExtractor={(player) => player.id.toString()}
      onEndReached={() => {
        setOnEndReachedState(true);
        onEndReached();
      }}
    />
  );
};

const styles = StyleSheet.create({
  container: {
    paddingHorizontal: 16,
  },
  playerName: {
    fontWeight: 'bold',
  },
  playerAvatarContainer: {
    justifyContent: 'center',
    marginRight: 4,
  },
  rightIndexStyle: {
    fontWeight: 'bold',
    textAlign: 'justify',
    marginRight: 12,
    paddingTop: 6,
  },
});
