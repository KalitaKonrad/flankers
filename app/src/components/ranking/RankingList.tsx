import React, { ReactComponentElement, useState } from 'react';
import { FlatList, ListRenderItemInfo, StyleSheet, View } from 'react-native';
import { List, Text, useTheme } from 'react-native-paper';

import { TeamProfilePayload } from '../../types/teamProfile';
import { UserProfilePayload } from '../../types/userProfilePayload';
import { Avatar } from '../shared/Avatar';

interface PlayersRankingProps {
  data: TeamProfilePayload[] | UserProfilePayload[];
  pageNumber: number;
  userId: number;
  buttonGroup: React.ComponentType<any>;
}

export const RankingList: React.FC<PlayersRankingProps> = ({
  data,
  pageNumber,
  userId,
  buttonGroup,
}) => {
  const theme = useTheme();

  const renderItem = ({
    item,
    index,
  }: ListRenderItemInfo<TeamProfilePayload | UserProfilePayload>) => (
    <List.Item
      title={item.name}
      titleStyle={styles.playerName}
      description={`${item.elo} punktÃ³w rankingowych`}
      left={() => (
        <View style={styles.playerAvatarContainer}>
          <Avatar size={40} borderRadius={8} src={{ uri: item.avatar }} />
        </View>
      )}
      right={() => (
        <Text style={styles.rightIndexStyle}>
          #{index + (pageNumber - 1) * 10 + 1}
        </Text>
      )}
      style={
        item.id === userId
          ? {
              backgroundColor: theme.colors.primary,
              borderRadius: 8,
              opacity: 0.5,
            }
          : { backgroundColor: theme.colors.white }
      }
    />
  );

  return (
    <FlatList
      data={data as any}
      renderItem={renderItem}
      contentContainerStyle={styles.container}
      keyExtractor={(player) => player.id.toString()}
      ListFooterComponent={buttonGroup}
    />
  );
};

const styles = StyleSheet.create({
  container: {
    paddingHorizontal: 16,
  },
  playerName: {
    fontWeight: 'bold',
  },
  playerAvatarContainer: {
    justifyContent: 'center',
    marginRight: 4,
  },
  rightIndexStyle: {
    fontWeight: 'bold',
    textAlign: 'justify',
    marginRight: 12,
    paddingTop: 6,
  },
});
