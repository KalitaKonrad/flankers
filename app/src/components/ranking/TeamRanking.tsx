import React from 'react';
import { FlatList, ListRenderItemInfo, StyleSheet, View } from 'react-native';
import { List, Text } from 'react-native-paper';

import { useTeamLeaderboardsQuery } from '../../hooks/useTeamLeaderboardsQuery';
import { TeamProfilePayload } from '../../types/teamProfile';
import { Avatar } from '../shared/Avatar';

interface PlayersRankingProps {
  teams: TeamProfilePayload[];
}

export const TeamsRanking: React.FC<PlayersRankingProps> = ({ teams }) => {
  const renderItem = ({
    item,
    index,
  }: ListRenderItemInfo<TeamProfilePayload>) => (
    <List.Item
      title={item.name}
      titleStyle={styles.playerName}
      description={`${item.elo} punktÃ³w rankingowych`}
      left={() => (
        <View style={styles.playerAvatarContainer}>
          <Avatar size={40} borderRadius={8} src={{ uri: item.avatar }} />
        </View>
      )}
      right={() => <Text style={styles.rightIndexStyle}>{index + 1}</Text>}
    />
  );

  return (
    <FlatList
      data={teams}
      renderItem={renderItem}
      contentContainerStyle={styles.container}
      keyExtractor={(player) => player.id.toString()}
    />
  );
};

const styles = StyleSheet.create({
  container: {
    paddingHorizontal: 16,
  },
  playerName: {
    fontWeight: 'bold',
  },
  playerAvatarContainer: {
    justifyContent: 'center',
    marginRight: 4,
  },
  rightIndexStyle: {
    fontWeight: 'bold',
    textAlign: 'justify',
    marginRight: 12,
    paddingTop: 6,
  },
});
