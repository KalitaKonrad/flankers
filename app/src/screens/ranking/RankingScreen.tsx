import { StackScreenProps } from '@react-navigation/stack';
import React, { useEffect, useState } from 'react';
import { StyleSheet, View } from 'react-native';

import { Container } from '../../components/layout/Container';
import { RankingList } from '../../components/ranking/RankingList';
import { AppButton } from '../../components/shared/AppButton';
import { Switch } from '../../components/shared/Switch';
import { useAlert } from '../../hooks/useAlert';
import { usePlayerLeaderboardsQuery } from '../../hooks/usePlayerLeaderboardsQuery';
import { useTeamLeaderboardsQuery } from '../../hooks/useTeamLeaderboardsQuery';
import { useUserProfileQuery } from '../../hooks/useUserProfileQuery';
import { ListPlaceholder } from '../../utils/ListPlaceholder';
import { RankingScreenStackParamList } from './RankingScreenStack';

type RankingScreenProps = StackScreenProps<
  RankingScreenStackParamList,
  'Ranking'
>;

export const RankingScreen: React.FC<RankingScreenProps> = ({ navigation }) => {
  const [showTeamsRanking, setShowTeamsRanking] = useState(false);

  const [pagePlayerRanking, setPagePlayerRanking] = useState(1);
  const [pageTeamRanking, setPageTeamRanking] = useState(1);

  const profile = useUserProfileQuery();
  const { showAlert } = useAlert();

  const teamLeaderboardsQuery = useTeamLeaderboardsQuery(pageTeamRanking);

  const playerLeaderboardsQuery = usePlayerLeaderboardsQuery(pagePlayerRanking);

  const onNextPagePlayerRanking = () => {
    if (
      !playerLeaderboardsQuery.isPreviousData &&
      playerLeaderboardsQuery.data?.next_page_url
    ) {
      setPagePlayerRanking(pagePlayerRanking + 1);
    }
  };

  const onNextPageTeamRanking = () => {
    if (
      !teamLeaderboardsQuery.isPreviousData &&
      teamLeaderboardsQuery.data?.next_page_url
    ) {
      setPageTeamRanking(pageTeamRanking + 1);
    }
  };

  const onPreviousPagePlayerRanking = () => {
    const value = Math.max(1, pagePlayerRanking - 1);
    setPagePlayerRanking(value);
  };

  const onPreviousPageTeamRanking = () => {
    const value = Math.max(1, pageTeamRanking - 1);
    setPageTeamRanking(value);
  };

  useEffect(() => {
    if (playerLeaderboardsQuery.isError || teamLeaderboardsQuery.isError) {
      showAlert('Ups', 'Błąd podczas wyświetlania rankingu');
    }
  }, [
    playerLeaderboardsQuery.isError,
    showAlert,
    teamLeaderboardsQuery.isError,
  ]);

  return (
    <Container>
      <View style={styles.switchContainer}>
        <Switch
          leftLabel="Gracze"
          rightLabel="Drużyny"
          onSwitchToLeft={() => setShowTeamsRanking(false)}
          onSwitchToRight={() => setShowTeamsRanking(true)}
          disabled={profile.isFetching}
        />
      </View>

      {playerLeaderboardsQuery.isLoading ? (
        <ListPlaceholder placeholderCount={6} itemHeight={50} />
      ) : showTeamsRanking ||
        playerLeaderboardsQuery.data === undefined ||
        profile.data?.id === undefined ? null : (
        <RankingList
          data={playerLeaderboardsQuery.data.data}
          pageNumber={pagePlayerRanking}
          userId={profile.data?.id}
          buttonGroup={
            <View style={styles.buttonGroup}>
              <AppButton
                onPress={onPreviousPagePlayerRanking}
                disabled={pagePlayerRanking === 1}>
                Wstecz
              </AppButton>
              <AppButton
                onPress={onNextPagePlayerRanking}
                disabled={
                  playerLeaderboardsQuery.isPreviousData ||
                  playerLeaderboardsQuery.data.next_page_url === null
                }>
                Naprzód
              </AppButton>
            </View>
          }
        />
      )}

      {teamLeaderboardsQuery.isLoading && showTeamsRanking ? (
        <ListPlaceholder placeholderCount={6} itemHeight={50} />
      ) : !showTeamsRanking ||
        teamLeaderboardsQuery.data === undefined ||
        profile.data?.current_team_id === undefined ? null : (
        <RankingList
          data={teamLeaderboardsQuery.data.data}
          pageNumber={pageTeamRanking}
          userTeamId={profile.data.current_team_id}
          buttonGroup={
            <View style={styles.buttonGroup}>
              <AppButton
                onPress={onPreviousPageTeamRanking}
                disabled={pageTeamRanking === 1}>
                Wstecz
              </AppButton>
              <AppButton
                onPress={onNextPageTeamRanking}
                disabled={
                  teamLeaderboardsQuery.isPreviousData ||
                  teamLeaderboardsQuery.data.next_page_url === null
                }>
                Naprzód
              </AppButton>
            </View>
          }
        />
      )}
    </Container>
  );
};

const styles = StyleSheet.create({
  switchContainer: {
    paddingHorizontal: 16,
    marginVertical: 24,
  },
  buttonGroup: {
    flexDirection: 'row',
    justifyContent: 'space-around',
    marginBottom: 16,
  },
});
