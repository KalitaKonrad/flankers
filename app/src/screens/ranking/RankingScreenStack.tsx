import { MaterialBottomTabScreenProps } from '@react-navigation/material-bottom-tabs';
import { createStackNavigator } from '@react-navigation/stack';
import React from 'react';
import { useTheme } from 'react-native-paper';

import { useNotificationHandler } from '../../hooks/useNotificationHandler';
import { BottomTabNavigationParamList } from '../AppScreenStack';
import { RankingScreen } from './RankingScreen';

export type RankingScreenStackParamList = {
  Ranking: undefined;
};

type RankingScreenStackProps = MaterialBottomTabScreenProps<
  BottomTabNavigationParamList,
  'Ranking'
>;

const Stack = createStackNavigator<RankingScreenStackParamList>();

export const RankingScreenStack: React.FC<RankingScreenStackProps> = ({
  navigation,
}) => {
  const theme = useTheme();
  useNotificationHandler(navigation);

  return (
    <Stack.Navigator
      initialRouteName="Ranking"
      screenOptions={theme.primaryHeader}>
      <Stack.Screen
        name="Ranking"
        component={RankingScreen}
        options={{ title: 'Ranking' }}
      />
    </Stack.Navigator>
  );
};
