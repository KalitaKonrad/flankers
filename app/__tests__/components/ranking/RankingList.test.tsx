import { render } from '@testing-library/react-native';
import * as React from 'react';
import { Button } from 'react-native';

import { RankingList } from '../../../src/components/ranking/RankingList';

describe('<RankingList>', () => {
  it('renders correctly when empty array with data provided', () => {
    const tree = render(
      <RankingList
        buttonGroup={<Button title="Hehe" onPress={() => null} />}
        data={[]}
        pageNumber={1}
      />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('renders correctly when all data provided', () => {
    const mockData = [
      {
        id: 5,
        name: 'hehe',
        description: 'string',
        owner_id: 5,
        updated_at: 'string',
        created_at: 'string',
        elo: 5,
        versioned_avatar: 'string',
      },
    ];

    const tree = render(
      <RankingList
        buttonGroup={<Button title="Hehe" onPress={() => null} />}
        data={mockData}
        pageNumber={2}
      />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
