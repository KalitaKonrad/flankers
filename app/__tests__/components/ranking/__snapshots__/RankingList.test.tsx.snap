// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RankingList> renders correctly when all data provided 1`] = `
<RCTScrollView
  ListFooterComponent={
    <Button
      onPress={[Function]}
      title="Hehe"
    />
  }
  contentContainerStyle={
    Object {
      "paddingHorizontal": 16,
    }
  }
  data={
    Array [
      Object {
        "created_at": "string",
        "description": "string",
        "elo": 5,
        "id": 5,
        "name": "hehe",
        "owner_id": 5,
        "updated_at": "string",
        "versioned_avatar": "string",
      },
    ]
  }
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  updateCellsBatchingPeriod={50}
  viewabilityConfigCallbackPairs={Array []}
  windowSize={21}
>
  <View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        accessible={true}
        focusable={false}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Array [
            false,
            Array [
              Object {
                "padding": 8,
              },
              Object {
                "backgroundColor": undefined,
              },
            ],
          ]
        }
      >
        <View
          style={
            Object {
              "flexDirection": "row",
            }
          }
        >
          <View
            style={
              Object {
                "justifyContent": "center",
                "marginRight": 4,
              }
            }
          >
            <View
              style={
                Array [
                  Object {
                    "backgroundColor": "#fff",
                  },
                  Object {
                    "borderRadius": 8,
                    "elevation": 0,
                    "height": 40,
                    "width": 40,
                  },
                  Object {
                    "shadowColor": "rgba(0, 0, 0, 0.1)",
                    "shadowOffset": Object {
                      "height": 4,
                      "width": 0,
                    },
                    "shadowOpacity": 1,
                    "shadowRadius": 4,
                  },
                  undefined,
                ]
              }
            >
              <Image
                source={
                  Object {
                    "uri": "string",
                  }
                }
                style={
                  Array [
                    Object {
                      "borderColor": "#FFFFFF",
                    },
                    Object {
                      "borderRadius": 8,
                      "borderWidth": 0,
                      "height": 40,
                      "width": 40,
                    },
                  ]
                }
              />
            </View>
          </View>
          <View
            style={
              Array [
                Object {
                  "marginVertical": 6,
                  "paddingLeft": 8,
                },
                Object {
                  "flex": 1,
                  "justifyContent": "center",
                },
              ]
            }
          >
            <Text
              numberOfLines={1}
              selectable={false}
              style={
                Array [
                  Object {
                    "color": "#000000",
                    "fontFamily": "System",
                    "fontWeight": "400",
                    "textAlign": "left",
                  },
                  Array [
                    Object {
                      "fontSize": 16,
                    },
                    Object {
                      "color": "rgba(0, 0, 0, 0.87)",
                    },
                    Object {
                      "fontWeight": "bold",
                    },
                  ],
                ]
              }
            >
              hehe
            </Text>
            <Text
              numberOfLines={2}
              selectable={false}
              style={
                Array [
                  Object {
                    "color": "#000000",
                    "fontFamily": "System",
                    "fontWeight": "400",
                    "textAlign": "left",
                  },
                  Array [
                    Object {
                      "fontSize": 14,
                    },
                    Object {
                      "color": "rgba(0, 0, 0, 0.54)",
                    },
                    undefined,
                  ],
                ]
              }
            >
              5 punkt√≥w rankingowych
            </Text>
          </View>
          <Text
            style={
              Array [
                Object {
                  "fontSize": 14,
                  "fontWeight": "normal",
                },
                Object {
                  "fontWeight": "bold",
                  "marginRight": 12,
                  "paddingTop": 6,
                  "textAlign": "justify",
                },
              ]
            }
          >
            #
            11
          </Text>
        </View>
      </View>
    </View>
    <View
      onLayout={[Function]}
    >
      <View
        accessibilityRole="button"
        accessibilityState={Object {}}
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Array [
              Object {},
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#007AFF",
                  "fontSize": 18,
                  "margin": 8,
                  "textAlign": "center",
                },
              ]
            }
          >
            Hehe
          </Text>
        </View>
      </View>
    </View>
  </View>
</RCTScrollView>
`;

exports[`<RankingList> renders correctly when empty array with data provided 1`] = `
<RCTScrollView
  ListFooterComponent={
    <Button
      onPress={[Function]}
      title="Hehe"
    />
  }
  contentContainerStyle={
    Object {
      "paddingHorizontal": 16,
    }
  }
  data={Array []}
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  updateCellsBatchingPeriod={50}
  viewabilityConfigCallbackPairs={Array []}
  windowSize={21}
>
  <View>
    <View
      onLayout={[Function]}
    >
      <View
        accessibilityRole="button"
        accessibilityState={Object {}}
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Array [
              Object {},
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#007AFF",
                  "fontSize": 18,
                  "margin": 8,
                  "textAlign": "center",
                },
              ]
            }
          >
            Hehe
          </Text>
        </View>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
